forest/willow/log_stats.py

```diff
@@ -140,6 +140,69 @@ def text_analysis(
     )
 
 
+def text_and_call_analysis(
+    df_call: pd.DataFrame, df_text: pd.DataFrame, stamp: int, step_size: int
+) -> tuple:
+    """Calculate the summary statistics for the call data
+    in the given time interval.
+
+    Args:
+        df_call: pd.DataFrame
+            dataframe of the call data
+        df_text: pd.DataFrame
+            dataframe of the text data
+        stamp: int
+            starting timestamp of the study
+        step_size: int
+            ending timestamp of the study
+
+    Returns:
+        tuple of summary statistics containing:
+            num_uniq_individuals_call_or_text: int
+                number of people making incoming calls or texts to the Beiwe
+                user or who the Beiwe user made outgoing calls or texts to
+
+
+    """
+    # filter the data based on the timestamp
+    if df_call.shape[0] > 0:
+        temp_call = df_call[
+            (df_call["timestamp"] / 1000 >= stamp)
+            & (df_call["timestamp"] / 1000 < stamp + step_size)
+        ]
+        index_in_call = np.array(temp_call["call type"]) == "Incoming Call"
+        index_out_call = np.array(temp_call["call type"]) == "Outgoing Call"
+        index_mis_call = np.array(temp_call["call type"]) == "Missed Call"
+        calls_in = np.array(temp_call["hashed phone number"])[index_in_call]
+        calls_out = np.array(temp_call["hashed phone number"])[index_out_call]
+        calls_mis = np.array(temp_call["hashed phone number"])[index_mis_call]
+
+    else:  # no calls were received, so no unique numbers will be used
+        calls_in = np.array([])
+        calls_out = np.array([])
+
+    if df_text.shape[0] > 0:
+        temp_text = df_text[
+            (df_text["timestamp"] / 1000 >= stamp)
+            & (df_text["timestamp"] / 1000 < stamp + step_size)
+        ]
+
+        index_s = np.array(temp_text["sent vs received"]) == "sent SMS"
+        index_r = np.array(temp_text["sent vs received"]) == "received SMS"
+        texts_in = np.array(temp_text["hashed phone number"])[index_r]
+        texts_out = np.array(temp_text["hashed phone number"])[index_s]
+    else:  # no texts were received, so no unique numbers will be used
+        texts_in = np.array([])
+        texts_out = np.array([])
+
+    num_uniq_individuals_call_or_text = len(np.unique(np.hstack(
+        [calls_in, texts_in, texts_out, calls_out, calls_mis]
+    )))
+    return (
+        num_uniq_individuals_call_or_text,
+    )
```

Comment:
Why are we returning a tuple here and not an integer?

Reply:
I guess just for forward compatibility. I thought that it might be possible that we would want to see the number of unique people with incoming calls or text, or the number of people with outgoing calls or texts

Reply:
ok sure!